ddtrace/contrib/internal/vertexai/_utils.py

```diff
@@ -0,0 +1,312 @@
+import copy
+import sys
+
+from vertexai.generative_models import GenerativeModel
+from vertexai.generative_models import Part
+
+from ddtrace.internal.utils import get_argument_value
+from ddtrace.llmobs._utils import _get_attr
+
+
+class BaseTracedVertexAIStreamResponse:
+    def __init__(self, generator, integration, span):
+        self._generator = generator
+        self._dd_integration = integration
+        self._dd_span = span
+        self._chunks = []
+
+
+class TracedVertexAIStreamResponse(BaseTracedVertexAIStreamResponse):
+    def __enter__(self):
+        self._generator.__enter__()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._generator.__exit__(exc_type, exc_val, exc_tb)
+
+    def __iter__(self):
+        try:
+            for chunk in self._generator.__iter__():
+                self._chunks.append(copy.deepcopy(chunk))
+                yield chunk
+        except Exception:
+            self._dd_span.set_exc_info(*sys.exc_info())
+            raise
+        else:
+            tag_stream_response(self._dd_span, self._chunks, self._dd_integration)
+        self._dd_span.finish()
+
+
+class TracedAsyncVertexAIStreamResponse(BaseTracedVertexAIStreamResponse):
+    def __aenter__(self):
+        self._generator.__enter__()
+        return self
+
+    def __aexit__(self, exc_type, exc_val, exc_tb):
+        self._generator.__exit__(exc_type, exc_val, exc_tb)
+
+    async def __aiter__(self):
+        try:
+            async for chunk in self._generator.__aiter__():
+                self._chunks.append(copy.deepcopy(chunk))
+                yield chunk
+        except Exception:
+            self._dd_span.set_exc_info(*sys.exc_info())
+            raise
+        else:
+            tag_stream_response(self._dd_span, self._chunks, self._dd_integration)
+        self._dd_span.finish()
```

Comment:
oh after reading more I see that `finish()` is called on error in the wrapper method (under the `if span.error or not stream:` condition) so the `finally` is not required here, although it would be worth keeping to avoid a regression if the outer code is changed.

Reply:
sounds good!